STAC-13165 and STAC-13066 Fix missing http stats / Add ipv6 support.

LouisParkin opened the pull request:
[don't merge] debugging

[don't merge] debugging [1]

[don't merge] debugging [2]

[don't merge] debugging [3]

[don't merge] debugging [4]

[don't merge] debugging [5]

[don't merge] debugging [7]

STAC-13165 Fix missing http stats.

[don't merge] debugging [7.1]

[don't merge] debugging [7.2]


[don't merge] debugging [7.3]


[don't merge] debugging [7.4]


[don't merge] debugging [7.5]


[don't merge] debugging [7.6]


[don't merge] debugging [7.7]


[don't merge] debugging [7.8]


[don't merge] debugging [7.9]


[don't merge] debugging [7.10]


[don't merge] debugging [7.11]


[don't merge] debugging [7.12]


[don't merge] debugging [7.13]


[don't merge] debugging [7.14]


[don't merge] debugging [7.15]


[don't merge] debugging [7.16]


[don't merge] testing [7.17]


[don't merge] testing [7.18]


[don't merge] testing [7.19]

Q69K commented on cmd/nettop/nettop.go:
```diff
@@ -60,8 +62,8 @@ func main() {
-			case now := <-tick.C:
-				printConns(now)
+			//case now := <-tick.C:
+			//	printConns(now)
```
i'd revert this

LouisParkin replied:
Ok, done.

Q69K commented on tcptracer-bpf.h:
```diff
@@ -91,10 +93,23 @@ struct event_mysql_greeting {
+struct event_http_response_v6 {
+    __u16 status_code;
+    __u32 response_time;
+    struct ipv6_tuple_t connection;
```
looks like it's better to move connection to `struct perf_event` / `struct perf_event_v6` . Otherwise we gonna create two structures for every kind of event.

LouisParkin replied:
@Q69K how about if I replaced the v4/v6 distinction with this:
```
enum connection_types {
    CONNECTION_TYPE_IPv4 = 0,
    CONNECTION_TYPE_IPv6 = 1,
    CONNECTION_TYPE_MAX
};

union un_connections {
    struct ipv4_tuple_t ipv4_connection;
    struct ipv6_tuple_t ipv6_connection;
};

struct connection_wrapper {
    connection_types connection_type;
    un_connections connection;
};

struct event_http_response {
    __u16 status_code;
    __u32 response_time;
    struct connection_wrapper
};

struct event_mysql_greeting {
    __u16 protocol_version;
    struct connection_wrapper
};
```

LouisParkin replied:
Then the :arrow_up: `event_http_response` and `event_mysql_greeting` structures can be used for ipv4 and ipv6...

Q69K replied:
That's good!

Just it was initially modelled a bit wrong: http has no idea that there are some IPs, so I'd keep onion like 

```
connection_event {
   address tuple
   union {
       http response  {
           status code
           resp time
       }
       mysql greeting  {
           protocol version
      }
}
```

But anyway is good until it leads to unnecessary duplicates

LouisParkin replied:
Done

Q69K commented on tcptracer-bpf.c:
```diff
@@ -780,8 +781,8 @@ __attribute__((always_inline))
 	if (size > 11) {
-		if (memcmp(buffer, http_marker, sizeof(http_marker)) == 0) {
-			int status_code = 100 * (buffer[9] - '0') + 10 * (buffer[10] - '0') + (buffer[11] - '0');
+		if (memcmp(buffer, http_marker, 4) == 0) {
```
i'd keep sizeof()

LouisParkin replied:
That was debug code :fearful: I'll change it back!

Q69K commented on tcptracer-bpf.c:
```diff
@@ -854,42 +939,54 @@ static int tcp_send(struct pt_regs *ctx, const size_t size) {
 		return 0;
 	}
 
-	if (status->protocol_inspection_enabled) {
```
feature flag is lost

LouisParkin replied:
I will put it back.

Q69K commented on tcptracer-bpf.c:
```diff
@@ -854,42 +939,54 @@ static int tcp_send(struct pt_regs *ctx, const size_t size) {
+
+                if (parse_http_response(data, iov.iov_len, &http_status_code)) {
+                    send_http_response_v6(ctx, t6, http_status_code, ttfb / 1000, current_time, cpu);
+                } else {
```
`else if (parse_mysql_greeting(data, iov.iov_len, &mysql_greeting_protocol_version))` was removed only for debugging purposes, it should be here, otherwise we gonna mark every other connection as mysql! :D

LouisParkin replied:
Handled it (this is probably part of the debug you were doing before you and @julianosk jumped on that other ticket, so I missed it :frowning_face: )

LouisParkin replied:
Put it back

Q69K commented on pkg/tracer/tracer_linux.go:
```diff
@@ -610,9 +610,52 @@ func (t *LinuxTracer) dispatchPerfEvent(event *common.PerfEvent) {
+		}
+		conn.ApplicationProtocol = "mysql"
+		t.tcpConnInsights[mysqlGreeting.Connection] = conn
+	} else if event.HTTPResponseV6 != nil {
```
also that structure of event leads to all this copies of the same code

LouisParkin replied:
Agreed, will see how we can do this better.

LouisParkin replied:
Done

Q69K commented on tcptracer-bpf.h:
```diff
@@ -79,16 +79,23 @@ struct tracked_socket {
+#define EVENT_HTTP_RESPONSE_V6 3
+#define EVENT_MYSQL_GREETING_V6 4
+
+struct connections {
```
shouldnt it be a union?

LouisParkin replied:
It was, but that complicated the golang implementation a bit.  I can take another stab at it, I think I understand it better now. Should I @Q69K ?

LouisParkin replied:
Did it anyway.

Q69K commented on pkg/tracer/tracer_linux.go:
```diff
@@ -579,12 +579,18 @@ func (t *LinuxTracer) getMap(mapName string) (*bpflib.Map, error) {
 	if event.HTTPResponse != nil {
-		logger.Tracef("http response: %v", event.HTTPResponse)
+		if event.HTTPResponse.Connection.IPV4Connection.Lport == 0 {
+			logger.Warnf("http response ipv6: %v", event.HTTPResponse.Connection.IPV6Connection)
```
probably, not "warning", but "trace" or "debug"

LouisParkin replied:
This was your code :laughing: . I'll change to trace.

Q69K commented on pkg/tracer/tracer_linux.go:
```diff
@@ -579,12 +579,18 @@ func (t *LinuxTracer) getMap(mapName string) (*bpflib.Map, error) {
+			logger.Warnf("http response ipv6: %v", event.HTTPResponse.Connection.IPV6Connection)
+			connection = event.HTTPResponse.Connection.IPV6Connection
+		} else {
+			logger.Warnf("http response ipv4: %v", event.HTTPResponse.Connection.IPV4Connection)
```
here as well

julianosk commented on pkg/tracer/common/model.go:
```diff
@@ -156,28 +156,33 @@ func (m *Histogram) MarshalJSON() ([]byte, error) {
 }
 
-type ConnTupleV4 struct {
+type ConnTuple struct {
```
Should we create another Tuple, like the C code:
```
struct ipv6_tuple_t {
	/* Using the type unsigned __int128 generates an error in the ebpf verifier */
	__u64 laddr_h;
	__u64 laddr_l;
	__u64 raddr_h;
	__u64 raddr_l;
	__u16 lport;
	__u16 rport;
	__u32 netns;
	__u32 pid;
};
```

LouisParkin replied:
Not needed, the golang code is only interested in Laddr, Raddr, Lport, Rport, and Pid.

julianosk commented on tcptracer-bpf.h:
```diff
@@ -79,16 +79,23 @@ struct tracked_socket {
+#define EVENT_HTTP_RESPONSE_V6 3
+#define EVENT_MYSQL_GREETING_V6 4
+
+union connections {
```
Maybe it makes more sense to move the type of the connection to a `connection_wrapper` or something like this. Then you have only one type of `EVENT_HTTP_RESPONSE` and `EVENT_MYSQL_GREETING` and you don't have to repeat a lot of code.

julianosk commented on tcptracer-bpf.h:
```diff
@@ -102,6 +107,7 @@ struct perf_event
 	__u16 event_type;
 	__u64 timestamp;
 	union event_payload payload;
+	union connections connection;
```
@LouisParkin It looks nicer. Maybe we could also have a `connection_type` here. Then we wouldn't need to have 2 event types for each application protocol. What do you think?

LouisParkin replied:
The effect on the golang code is the same.  There has to be an `if` or a `switch-case` to decided how we extract the information.  If it is a perf event field indicating network protocol separate from application protocol, I agree, in principle it is the cleaner option.

Q69K commented on pkg/tracer/tracer_linux.go:
```diff
@@ -608,21 +609,20 @@ func (t *LinuxTracer) dispatchPerfEvent(event *common.PerfEvent) {
 	} else if event.MySQLGreeting != nil {
-		logger.Tracef("mysql greeting: %v", event.MySQLGreeting)
-		mysqlGreeting := event.MySQLGreeting
+		logger.Warnf("mysql greeting: %v", event.Connection)
```
also trace

LouisParkin replied:
This was for debugging, it has been reverted to trace in the latest commit

Q69K commented on pkg/tracer/tracer_linux.go:
```diff
@@ -579,12 +579,13 @@ func (t *LinuxTracer) getMap(mapName string) (*bpflib.Map, error) {
+	connection := *event.Connection
 	if event.HTTPResponse != nil {
-		logger.Tracef("http response: %v", event.HTTPResponse)
+		logger.Warnf("http response: %v", event.Connection)
```
trace

LouisParkin replied:
This was for debugging, it has been reverted to trace in the latest commit

julianosk commented on tcptracer-bpf.c:
```diff
@@ -30,6 +30,7 @@
 		bpf_trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
 	})
 
+#define TO_HEX(i) (i <= 9 ? '0' + i : 'A' - 10 + i)
```
I couldn't find usages of this macro.

LouisParkin replied:
That was debug code I added. Will remove.

Q69K commented on tcptracer-bpf.c:
```diff
@@ -30,6 +30,7 @@
 		bpf_trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
 	})
 
+#define TO_HEX(i) (i <= 9 ? '0' + i : 'A' - 10 + i)
```
```suggestion
```

julianosk commented on tcptracer-bpf.c:
```diff
@@ -854,43 +896,62 @@ static int tcp_send(struct pt_regs *ctx, const size_t size) {
+                struct ipv4_tuple_t t = {};
+                struct ipv6_tuple_t t6 = {};
+
+                if(!is_v4_or_v6(sk, status)){
```
It looks this check can be moved upper.

LouisParkin replied:
Good catch.

LouisParkin replied:
Good catch.

julianosk commented on tcptracer-bpf.c:
```diff
@@ -854,43 +896,59 @@ static int tcp_send(struct pt_regs *ctx, const size_t size) {
+    if (status->protocol_inspection_enabled) {
+        struct msghdr msg = {};
+        bpf_probe_read(&msg, sizeof(msg), k_msg);
+    //	bpf_debug("test %d <> %d\n", msg.msg_iter.type & ~(READ | WRITE), status->iter_type);
```
```suggestion
```

julianosk commented on tcptracer-bpf.c:
```diff
@@ -854,43 +896,59 @@ static int tcp_send(struct pt_regs *ctx, const size_t size) {
+        if ((msg.msg_iter.type & ~(READ | WRITE)) == status->iter_type) {
+            char *data = bpf_map_lookup_elem(&write_buffer_heap, &zero);
+            if (data != NULL) {
+                if(!is_v4_or_v6(sk, status)){
```
I would move it to just after the flag check (`if (status->protocol_inspection_enabled) {`). 
Another thing, should we just `return 0`? Shouldn't we `return increment_tcp_stats(sk, status, size, 0);`?

Q69K replied:
Oh, true! This whole block should not "return"

LouisParkin replied:
There was already a `return 0;` in the `tcp_send()` function.  Should I replace with `return increment_tcp_stats(sk, status, size, 0);`?  If it is not ip4 or ip6, do we call `return increment_tcp_stats(sk, status, size, 0);` or `return 0;`?

julianosk replied:
Maybe you could do on line 899:
```
    if (status->protocol_inspection_enabled && is_v4_or_v6(sk, status)) {
```

LouisParkin replied:
@julianosk we checked what `increment_tcp_stats(sk, status, size, 0);` actually does.  If not ip4 or ip6, it does nothing.  So we move the check to above the `if (status->protocol_inspection_enabled)` check, and it returns 0.

julianosk reviewed (commented):
Last few comments.. The only thing that bothers me a little bit is the indentation. The rest of the file is indented with tabs. Depending on your editor configuration, the code could look a bit messy.

LouisParkin commented:
> Last few comments.. The only thing that bothers me a little bit is the indentation. The rest of the file is indented with tabs. Depending on your editor configuration, the code could look a bit messy. What do you think?

I have no issue with following the tab convention.  History (and different compilers) have taught me that spaces are better than tabs, so my editors are usually set to use 4 spaces instead of a tab character.

julianosk commented:
> > Last few comments.. The only thing that bothers me a little bit is the indentation. The rest of the file is indented with tabs. Depending on your editor configuration, the code could look a bit messy. What do you think?
> 
> I have no issue with following the tab convention. History (and different compilers) have taught me that spaces are better than tabs, so my editors are usually set to use 4 spaces instead of a tab character.

I agree! I even thought of refactoring everything to use spaces when working on the previous story haha

LouisParkin commented:
> > > Last few comments.. The only thing that bothers me a little bit is the indentation. The rest of the file is indented with tabs. Depending on your editor configuration, the code could look a bit messy. What do you think?
> > 
> > 
> > I have no issue with following the tab convention. History (and different compilers) have taught me that spaces are better than tabs, so my editors are usually set to use 4 spaces instead of a tab character.
> 
> I agree! I even thought of refactoring everything to use spaces when working on the previous story haha

So @julianosk I changed all the 4-space groupings to tabs, and that broke the bpf code, so I changed it back, and decided to not mess with it. @Q69K says he is happy as long as it looks right in his IDE (only github/gitlab editor displays funny)

julianosk reviewed (approved):
LGTM
